protocol/contracts/zaps/GenericZap.sol

```diff
@@ -0,0 +1,697 @@
+
+    _approveToken(token1, address(uniswapV2Router), amountB);
+    _approveToken(token0, address(uniswapV2Router), amountA);
+
+    _addLiquidityUniV2(_pair, _for, amountA, amountB, _zapLiqRequest.uniAmountAMin, _zapLiqRequest.uniAmountBMin, _zapLiqRequest.shouldTransferResidual);
+  }
+
+  /**
+   * @dev Get minimum amounts fo token0 and token1 to tolerate when adding liquidity to uniswap v2 pair
+   * @param amountADesired Input desired amount of token0
+   * @param amountBDesired Input desired amount of token1
+   * @param pair Target uniswap v2 pair
+   * @return amountA Minimum amount of token0 to use when adding liquidity
+   * @return amountB Minimum amount of token1 to use when adding liquidity
+   */
+  function addLiquidityGetMinAmounts(
+    uint amountADesired,
+    uint amountBDesired,
+    IUniswapV2Pair pair
+  ) public view returns (uint amountA, uint amountB) {
+    (uint reserveA, uint reserveB,) = pair.getReserves();
+    if (reserveA == 0 && reserveB == 0) {
+      (amountA, amountB) = (amountADesired, amountBDesired);
+    } else {
+      uint amountBOptimal = Swap.quote(amountADesired, reserveA, reserveB);
+      if (amountBOptimal <= amountBDesired) {
+          //require(amountBOptimal >= amountBMin, 'TempleStableAMMRouter: INSUFFICIENT_STABLE');
+          (amountA, amountB) = (amountADesired, amountBOptimal);
+      } else {
+          uint amountAOptimal = Swap.quote(amountBDesired, reserveB, reserveA);
+          assert(amountAOptimal <= amountADesired);
+          //require(amountAOptimal >= amountAMin, 'TempleStableAMMRouter: INSUFFICIENT_TEMPLE');
+          (amountA, amountB) = (amountAOptimal, amountBDesired);
+      }
+    }
+  }
+
+  function _addLiquidityCurvePool(
+    address _pool,
+    address _recipient,
+    uint256 _minLiquidityOut,
+    bool _useAltFunction,
+    uint256[] memory _amounts
+  ) internal returns (uint256 liquidity) {
+    bool success;
+    bytes memory data;
+    if (_useAltFunction) {
+      //liquidity = ICurvePool(_pool).add_liquidity(_amounts, _minLiquidityOut, false, _recipient);
+      data = abi.encodeWithSelector(0x0b4c7e4d, _amounts, _minLiquidityOut, false, _recipient);
+      // reuse data
+      (success, data) = _pool.call{value:0}(data);
+    } else {
+      //liquidity = ICurvePool(_pool).add_liquidity(_amounts, _minLiquidityOut, _recipient);
+      data = abi.encodeWithSelector(0xad6d8c4a, _amounts, _minLiquidityOut, _recipient);
+      // reuse data
+      (success, data) = _pool.call{value:0}(data);
+    }
+    require(success, "Failed adding liquidity");
+    liquidity = abi.decode(data, (uint256));
+  }
+
+  function _addLiquidityUniV2(
+    address _pair,
+    address _recipient,
+    uint256 _amountA,
+    uint256 _amountB,
+    uint256 _amountAMin,
+    uint256 _amountBMin,
+    bool _shouldTransferResidual
+  ) internal {
+    address tokenA = IUniswapV2Pair(_pair).token0();
```

Comment:
I don't fully understand the stack-too-deep, but why not add tokenA/tokenB down too?

Reply:
it's because of the way variables are cycled. by scoping the rest of code, those variables are handled differently pertaining to the stack iirc.